components/raftstore/src/store/util.rs

```diff
@@ -1123,6 +1125,41 @@ pub fn check_conf_change(
     }
 }
 
+pub fn check_remove_node(
+    cfg: &Config,
+    change_peers: &[ChangePeerRequest],
+    peer_heartbeat: &collections::HashMap<u64, std::time::Instant>,
+) -> Result<()> {
+    // max heartbeats missed to qualify being a slow peer
+    const MAX_MISSED_HEARTBEATS: u32 = 8;
+    let mut slow_peer_count: u32 = 0;
+    for last_heartbeat in peer_heartbeat.values() {
+        if last_heartbeat.elapsed() > MAX_MISSED_HEARTBEATS * cfg.raft_heartbeat_interval() {
+            slow_peer_count += 1;
+        }
+    }
+
+    for cp in change_peers {
+        let (change_type, peer) = (cp.get_change_type(), cp.get_peer());
+        if change_type == ConfChangeType::RemoveNode {
```

Comment:
Should we consider `AddLearnerNode` too?

Reply:
There's no harm to AddLearnerNode.

Reply:
I think we should also forbid adding node in such case. Otherwise, it may block write until the new peer applies snapshot.

```rust
#[test]
fn test_add_node_on_partition() {
    let count = 4;
    let mut cluster = new_node_cluster(0, count);
    let pd_client = Arc::clone(&cluster.pd_client);
    // Disable default max peer number check.
    pd_client.disable_default_operator();
    cluster.cfg.raft_store.raft_heartbeat_ticks = 1;
    cluster.cfg.raft_store.raft_base_tick_interval = ReadableDuration::millis(10);
    let r1 = cluster.run_conf_change();

    cluster.must_put(b"k0", b"v0");
    pd_client.must_add_peer(r1, new_peer(2, 2));
    must_get_equal(&cluster.get_engine(2), b"k0", b"v0");
    pd_client.must_add_peer(r1, new_peer(3, 3));
    must_get_equal(&cluster.get_engine(3), b"k0", b"v0");

    // peer 3 isolation
    cluster.add_send_filter(IsolationFilterFactory::new(3));

    // Block peer (4, 4) at receiving snapshot.
    cluster.add_send_filter(CloneFilterFactory(
        RegionPacketFilter::new(r1, 4).msg_type(MessageType::MsgSnapshot),
    ));
    pd_client.must_add_peer(r1, new_peer(4, 4));
    cluster.must_put(b"k1", b"v1"); # This case panics here, timeout.
}
```

Reply:
@overvenus 
The failure here is because the new peer (peer4) is slow.  I think this is a valid concern. But this PR is to look at the existing peers' slowness by their last heartbeat time and prevent the remove node(s) if it's going to disrupt the quorum. 
For the new peer, we don't know if it's slow or not until the peer is added.  Maybe we should check the health of the new node before adding a peer, but it's another problem.

Reply:
@overvenus I got your point. Do you mean AddLearnerNode on existing peer (which means it's a demote). We need to check it as well indeed. Updated the code.